File: stacks_and_queues/valid_parentheses.js
/**
 * @param {string} s
 * @returns {boolean}
 */
var isValid = function(s) {
    let stack = [];
    const matching = {
        "(": ")",
        "[": "]",
        "{": "}"
    }

    for (const c of s)
    {
        // If c is an opening bracket
        if (c in matching)
        {
            stack.push(c);
        }
        else
        {
            if (!stack.length)
            {
                return false;
            }

            let previousOpening = stack.pop();
            if (matching[previousOpening] != c)
            {
                return false;
            }
        }
    }

    return !stack.length;
}

File: stacks_and_queues/valid_parentheses.py
def isValid(s: str) -> bool:
    stack = []
    matching = {"(": ")", "[": "]", "{": "}"}

    for c in s:
        # If c is an opening bracket
        if c in matching:
            stack.append(c)
        else:
            if not stack:
                return False
            
            previous_opening = stack.pop()
            if matching[previous_opening] != c:
                return False
            
    return not stack
